incubator/hnc/pkg/reconcilers/hierarchical_namespace.go

```diff
@@ -0,0 +1,49 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package reconcilers
+
+import (
+	"github.com/go-logr/logr"
+	api "github.com/kubernetes-sigs/multi-tenancy/incubator/hnc/api/v1alpha1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// HierarchicalNamespaceReconciler reconciles HierarchicalNamespace CRs to make sure
+// all the hierarchical namespaces are properly maintained.
+type HierarchicalNamespaceReconciler struct {
+	client.Client
+	Log logr.Logger
+}
+
+// Reconcile sets up some basic variables and then calls the business logic.
+// It currently only generates some logs for testing.
+func (r *HierarchicalNamespaceReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	if ex[req.Namespace] {
+		return ctrl.Result{}, nil
+	}
+
```

Comment:
Hmm. What does it mean if a user creates an HNS object in a namespace that should be excluded by HNC? Should we just ignore it?

Reply:
I think we can prevent it by webhook when users create it. We can still keep the exclusion here. I will add a comment here and a todo for webhook?

Or do you think we should add another status to HNS and report it here?

Reply:
I kind of like the idea of a status. I never like relying solely on webhooks :)

Reply:
smythe@ actually pointed out a downside of webhooks to me yesterday which I hadn't thought of (but fits my overall philosophy): imagine we make the list of excluded namespaces configurable one day. Then someone could exclude a namespace just as someone else creates an HNS. The webhook might pass, but by the time the object's _actually_ submitted, the NS is now excluded. Granted, this seems unlikely but it's an example of why we shouldn't assume that all our state is actually good.

Reply:
I added an "error" state to the CRD status and a todo comment here.